plutus-core/plutus-core/src/Crypto/BLS12_381/G2.hs

```diff
@@ -0,0 +1,107 @@
+-- editorconfig-checker-disable
+{-# LANGUAGE DeriveAnyClass        #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeApplications      #-}
+
+module Crypto.BLS12_381.G2
+    ( Element (..)
+    , add
+    , neg
+    , scalarMul
+    , hashToGroup
+    , compress
+    , uncompress
+    , zero
+    , memSizeBytes
+    , compressedSizeBytes
+    ) where
+
+import Crypto.External.EllipticCurve.BLS12_381 qualified as BlstBindings
+import Crypto.External.EllipticCurve.BLS12_381.Internal qualified as BlstBindings.Internal
+
+import Crypto.Utils (byteStringAsHex)
+import PlutusCore.Pretty.PrettyConst (ConstConfig)
+import Text.PrettyBy (PrettyBy, prettyBy)
+
+import Control.DeepSeq (NFData, rnf)
+import Data.Bifunctor (second)
+import Data.ByteString (ByteString, pack)
+import Data.Proxy (Proxy (..))
+import Flat
+import Prettyprinter
+
+{- | See Note [Wrapping the BLS12-381 types]. -}
+newtype Element = Element { unElement :: BlstBindings.Point2 }
+    deriving newtype (Eq)
+instance Show Element where
+    show  = byteStringAsHex . compress
+instance Pretty Element where
+    pretty = pretty . show
+instance PrettyBy ConstConfig Element where
+    prettyBy _ = pretty
+instance Flat Element where
+    decode = do
+        x <- decode
+        case BlstBindings.blsUncompress @BlstBindings.Curve2 x of
+             Left err -> fail $ show err
+             Right e  -> pure $ Element e
+    encode = encode . BlstBindings.blsCompress @BlstBindings.Curve2 . unElement
```

Comment:
I guess we did that for space reasons.  With compression G1 points take up 48 bytes and G2 points take up 96, but with serialisation it's 96 for G1 and 192 for G2, and some of the applications we were think about might require dozens of group points (signatures for ATMS, for example).  With the 16K script size limit you'd only be able to fit 85 G2 points into a script using serialisation (as opposed to compression) , and there'd be no room left for actual code.

Reply:
We _could_ give uses the choice of which serialisation format by adding more builtins, or maybe add boolean arguments to the `compress` and `uncompress` functions (which we'd probably want to rename) to tell them whether to use compression or not.  The question of what to do in `flat` would remain though.  Maybe we could choose a fixed format for direct `flat` serialisation and deserialisation, say the long format, and require users to compress to a bytestring and have that serialised in the normal way if they want to save space by using the short format.  That's all getting a bit complicated though.

It's worth saying something about how this works in `blst`.  All the deserialisation/uncompression functions get is a C pointer with no length information, so you have to be [careful](https://github.com/input-output-hk/cardano-base/blob/66662d497db74f99b78198c6bc510e597125b815/cardano-crypto-class/src/Cardano/Crypto/EllipticCurve/BLS12_381/Internal.hs#L709) about checking lengths in the Haskell interface.  If you just pass the contents of eg a two-byte Haskell bytestring directly to the C function, it'll read a whole 48/96/192 bytes regardless,  with possibly hilarious consequences.

Reply:
> However, I see a problem with costing of the function. It would no longer be constant.

Yes, that's right.  We'd need some kind of switch in the costing machinery to take account of which method we're using.  That'd be possible (there are already some costing functions that are split into multiple parts) but a bit annoying, so it might be simpler to add four extra built-in functions for the non-compact serialisation format. I guess there's a trade-off about how you deal with the complexity introduced by adding new functionality.